src/include/wiredtiger.in

```diff
@@ -3682,6 +3704,502 @@ struct __wt_extractor {
+	 *
+	 * This method is not called concurrently with other file system or
+	 * handle operations.
+	 *
+	 * @errors
+	 *
+	 * @param file_system the WT_FILE_SYSTEM
+	 * @param session the current WiredTiger session
+	 * @param from the original name of the object
+	 * @param to the new name for the object
+	 */
+	int (*rename)(WT_FILE_SYSTEM *file_system,
+	    WT_SESSION *session, const char *from, const char *to);
+
+	/*!
+	 * Return the size of a named file system object
+	 *
+	 * This method may be called concurrently with other file system or
+	 * handle operations.
+	 *
+	 * @errors
+	 *
+	 * @param file_system the WT_FILE_SYSTEM
+	 * @param session the current WiredTiger session
+	 * @param name the name of the file system object
+	 * @param[out] sizep the size of the file system entry
+	 */
+	int (*size)(WT_FILE_SYSTEM *file_system,
+	    WT_SESSION *session, const char *name, wt_off_t *sizep);
+
+	/*!
+	 * If non-NULL a callback performed when the file system is closed and
+	 * will no longer be used by the WiredTiger database.
+	 *
+	 * The WT_FILE_SYSTEM::terminate callback is intended to allow cleanup,
+	 * the handle will not be subsequently accessed by WiredTiger.
+	 */
+	int (*terminate)(WT_FILE_SYSTEM *file_system, WT_SESSION *session);
+};
+
+/*! WT_FILE_HANDLE::fadvise flags: no longer need */
+#define	WT_FILE_HANDLE_DONTNEED	1
+/*! WT_FILE_HANDLE::fadvise flags: will need */
+#define	WT_FILE_HANDLE_WILLNEED	2
+
+/*!
+ * A file handle implementation returned by WT_FILE_SYSTEM::open_file.
+ *
+ * See @ref custom_file_systems for more information.
+ *
+ */
+struct __wt_file_handle {
+	/*!
+	 * The name of the file.
+	 */
+	char *name;
+
+	/*!
+	 * Close a file handle; the WT_FILE_HANDLE will not be further accessed
+	 * by WiredTiger.
+	 *
+	 * This method is not called concurrently with other file system or
+	 * handle operations.
+	 *
+	 * @errors
+	 *
+	 * @param file_handle the WT_FILE_HANDLE
+	 * @param session the current WiredTiger session
+	 */
+	int (*close)(WT_FILE_HANDLE *file_handle, WT_SESSION *session);
+
+	/*!
+	 * Indicate expected future use of file ranges, based on the POSIX
+	 * 1003.1 standard fadvise.
+	 *
+	 * This method is not required, and should be set to NULL when not
+	 * supported by the file.
+	 *
+	 * This method may be called concurrently with other file system or
+	 * handle operations.
+	 *
+	 * @errors
+	 *
+	 * @param file_handle the WT_FILE_HANDLE
+	 * @param session the current WiredTiger session
+	 * @param offset the file offset
+	 * @param len the size of the advisory
+	 * @param advice one of ::WT_FILE_HANDLE_WILLNEED or
+	 *    ::WT_FILE_HANDLE_DONTNEED.
+	 */
+	int (*fadvise)(WT_FILE_HANDLE *file_handle,
+	    WT_SESSION *session, wt_off_t offset, wt_off_t len, int advice);
+
+	/*!
+	 * Ensure disk space is allocated for the file, based on the POSIX
+	 * 1003.1 standard fallocate.
+	 *
+	 * This method is not required, and should be set to NULL when not
+	 * supported by the file.
```

Comment:
What does it mean for `lock` not to work if the database is opened for updates?  I expected `lock` to be required for filesystems that support writes.